framework/captcha/CaptchaAction.php

```diff
@@ -246,121 +228,58 @@ protected function getSessionKey()
-        }
-        if ($imageLibrary === 'gd') {
-            return $this->renderImageByGD($code);
-        } elseif ($imageLibrary === 'imagick') {
-            return $this->renderImageByImagick($code);
-        } else {
-            throw new InvalidConfigException("Defined library '{$imageLibrary}' is not supported");
-        }
+        /* @var $factory DriverFactory */
+        $factory = Yii::createObject(DriverFactory::className());
+        $captchaDriver = $factory->make($this->imageLibrary);
+
+        return $captchaDriver->renderCaptcha($code, $this->_imageSettings);
     }
 
     /**
-     * Renders the CAPTCHA image based on the code using GD library.
-     * @param string $code the verification code
-     * @return string image contents in PNG format.
+     * Sets the HTTP headers needed by image response.
      */
-    protected function renderImageByGD($code)
+    protected function setHttpHeaders()
     {
-        $image = imagecreatetruecolor($this->width, $this->height);
-
-        $backColor = imagecolorallocate(
-            $image,
-            (int) ($this->backColor % 0x1000000 / 0x10000),
-            (int) ($this->backColor % 0x10000 / 0x100),
-            $this->backColor % 0x100
-        );
-        imagefilledrectangle($image, 0, 0, $this->width - 1, $this->height - 1, $backColor);
-        imagecolordeallocate($image, $backColor);
-
-        if ($this->transparent) {
-            imagecolortransparent($image, $backColor);
-        }
-
-        $foreColor = imagecolorallocate(
-            $image,
-            (int) ($this->foreColor % 0x1000000 / 0x10000),
-            (int) ($this->foreColor % 0x10000 / 0x100),
-            $this->foreColor % 0x100
-        );
-
-        $length = strlen($code);
-        $box = imagettfbbox(30, 0, $this->fontFile, $code);
-        $w = $box[4] - $box[0] + $this->offset * ($length - 1);
-        $h = $box[1] - $box[5];
-        $scale = min(($this->width - $this->padding * 2) / $w, ($this->height - $this->padding * 2) / $h);
-        $x = 10;
-        $y = round($this->height * 27 / 40);
-        for ($i = 0; $i < $length; ++$i) {
-            $fontSize = (int) (mt_rand(26, 32) * $scale * 0.8);
-            $angle = mt_rand(-10, 10);
-            $letter = $code[$i];
-            $box = imagettftext($image, $fontSize, $angle, $x, $y, $foreColor, $this->fontFile, $letter);
-            $x = $box[2] + $this->offset;
-        }
-
-        imagecolordeallocate($image, $foreColor);
-
-        ob_start();
-        imagepng($image);
-        imagedestroy($image);
+        Yii::$app->getResponse()->getHeaders()
+            ->set('Pragma', 'public')
+            ->set('Expires', '0')
+            ->set('Cache-Control', 'must-revalidate, post-check=0, pre-check=0')
+            ->set('Content-Transfer-Encoding', 'binary')
+            ->set('Content-type', 'image/png');
+    }
 
-        return ob_get_clean();
+    public function setWidth($width)
```

Comment:
Instead of wrapping `ImageSettings` properties into a setter/getter it is better to intorduce `imageSettings` property and force developer to use it for configuration like `DataProvider` operates `Sort` and `Pagination` objects.